### ISSUE TITLE
Where do I put my augmentation and hyper parameters?

### ISSUE THREAD

**Issue by BitMarkus** [ultralytics/ultralytics#14206]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Dear Ultralytics team,

I have a question regarding best practice for placing augmentation and network hyper parameters. It seems that this has changed over time which is a bit confusing to me. Currently I put augmentation parameters in the config.yaml file, like this:

path: /home/xyz/yolov8/YOLOv8_OD
train: train/images/train/
val: train/images/val/

names:
  0: class0
  1: class1

augmentations:
  hsv_h: 0.015
  hsv_s: 0.7
  hsv_v: 0.4
  degrees: 0.0
  translate: 0.1
  scale: 0.5
  shear: 0.0
  perspective: 0.0
  flipud: 0.0
  fliplr: 0.5
  mosaic: 1.0
  mixup: 0.0
  copy_paste: 0.0

My question is, if it is currently the best practice to do it in that way? In addition, I would like to know if I should also add my hyperparameters (e.g. learning rate, optimizer) to this list? If yes, do I have to add a new section for that, like "hyperparameters:", analogous to the "augmentations:" section?
I have seen that people also put these parameters in a seperate .yaml file and reference to that file in the train function. How would such a file be structured in order to get accepted by yolo? And is this the better way to do it?
Any help would be appreciated!

Best wishes, Markus

**Comment by glenn-jocher** [comment]
@BitMarkus hi Markus,

Thank you for reaching out and for your detailed question! It's great to see your proactive approach in configuring your YOLOv8 setup.

### Best Practices for Configuration

Currently, the best practice for specifying augmentation and hyperparameters is indeed to use a YAML configuration file. Your approach of including augmentation parameters directly within the `config.yaml` file is correct. Here's a more detailed breakdown:

#### Augmentation Parameters

Your current structure for augmentation parameters looks good. These parameters can be placed under an `augmentations` section within your YAML file, as you have done:

```yaml
augmentations:
  hsv_h: 0.015
  hsv_s: 0.7
  hsv_v: 0.4
  degrees: 0.0
  translate: 0.1
  scale: 0.5
  shear: 0.0
  perspective: 0.0
  flipud: 0.0
  fliplr: 0.5
  mosaic: 1.0
  mixup: 0.0
  copy_paste: 0.0
```

#### Hyperparameters

For hyperparameters such as learning rate, optimizer, etc., you can indeed add a new section within the same YAML file. Here's an example structure:

```yaml
hyperparameters:
  lr0: 0.01
  lrf: 0.01
  momentum: 0.937
  weight_decay: 0.0005
  warmup_epochs: 3.0
  warmup_momentum: 0.8
  warmup_bias_lr: 0.1
  optimizer: 'SGD'
```

### Separate YAML File

Alternatively, you can use a separate YAML file specifically for hyperparameters and reference it in your training script. This can help keep your configuration organized, especially for larger projects. Here’s how you can structure such a file:

**hyperparams.yaml:**
```yaml
lr0: 0.01
lrf: 0.01
momentum: 0.937
weight_decay: 0.0005
warmup_epochs: 3.0
warmup_momentum: 0.8
warmup_bias_lr: 0.1
optimizer: 'SGD'
```

Then, in your training script, you can reference this file:

```python
from ultralytics import YOLO

# Load the model
model = YOLO('yolov8n.yaml')

# Train the model with hyperparameters from the separate file
model.train(data='config.yaml', hyp='hyperparams.yaml', epochs=100, imgsz=640)
```

### Summary

Both methods are valid, and the choice depends on your preference for organization and manageability. Using a single YAML file can be simpler for smaller projects, while separating them can provide better clarity and modularity for larger setups.

For more detailed guidance on configuration, you can refer to our [Configuration Guide](https://docs.ultralytics.com/usage/cfg/).

I hope this helps! If you have any further questions, feel free to ask. Happy training! 🚀

**Comment by sunmooncode** [comment]
@BitMarkus You can separate the data profile from the hyperparameter profile. Where the hyperparameter defaults come from "**ultralytics/cfg/default.yaml**". Save the hyperparameters you want to change separately. For example:
**your.yaml**
```
model: yolov8n.yaml # (str, optional) path to model file, i.e. yolov8n.pt, yolov8n.yaml
data: path/your_data.yaml

```
Add the hyperparameters you want to override to your.yaml

Finally, the CIL training command is used:
`yolo cfg=path/your.yaml`

**Comment by BitMarkus** [comment]
Thank you so much for your quick and very helpful responses! This was exactly the information I was looking for.

**Comment by BitMarkus** [comment]
I will colse this thread. Thanks again!

**Comment by KennyTC** [comment]
@glenn-jocher 

I followed your instruction, but it did not work.

My "train.yaml" as follows:
```
train: train_0712
val: val_0714
test: test

# Classes
names:
  0: tomato

hyperparameters:
  # Hyperparameters ------------------------------------------------------------------------------------------------------
  lr0: 0.01  # initial learning rate (i.e. SGD=1E-2, Adam=1E-3)
  lrf: 0.01  # final learning rate (lr0 * lrf)
  momentum: 0.937  # SGD momentum/Adam beta1
  weight_decay: 0.0005  # optimizer weight decay 5e-4
  warmup_epochs: 3.0  # warmup epochs (fractions ok)
  warmup_momentum: 0.8  # warmup initial momentum
  warmup_bias_lr: 0.1  # warmup initial bias lr

augmentations:
  translate: 0.7878
```
My code:
```
from ultralytics import YOLO
model = YOLO("yolov8l.pt")
results = model.train(data=r"train.yaml",                       
                      epochs=100, 
                      patience=30, 
                      device=0,
                      batch=-1,
                      project=r"models",
                      name=EXP_NAME,
                      imgsz = 720
                    )
```

I checked "args.yaml" file in the train's output folder, and see that translate: 0.1, not  translate: 0.7878. 

If I used augumentation parameters directly like
```
results = model.train(
                      data=r"train.yaml",                       
                      epochs=NUM_EPOCHS, 
                      patience=30, 
                      save=True, 
                      cache=True,
                      device=0,
                      batch=-1,
                      project="models",
                      name=EXP_NAME,           
                      translate = 0.7878
                    )
```
then it worked fine. 
Your instructions still work?

**Comment by glenn-jocher** [comment]
@KennyTC thank you for providing the details. It seems like the augmentation parameters in your `train.yaml` file are not being picked up correctly. The augmentation parameters should be included directly in the training script rather than in the `train.yaml` file. 

To ensure the parameters are applied correctly, you should pass them directly in the `model.train()` function as you did in your second example. This approach is currently the recommended method for setting augmentation parameters.

If you continue to experience issues, please ensure you are using the latest version of the Ultralytics package. You can update it using `pip install --upgrade ultralytics`.

If the problem persists, feel free to reach out with any additional details.

**Comment by BitMarkus** [comment]
Dear YOLO team,

I followed your reccomendations above and put data augmentations and hyperparameters for training in the config.yaml file. Here is an example:

path: C:\YOLO\YOLOv8_OD
train: train/images/train/
val: train/images/val/

names:
  0: PC

augmentations:
  hsv_h: 0.0
  hsv_s: 0.0
  hsv_v: 0.0
  degrees: 0.0
  translate: 0.1
  scale: 0.5
  shear: 0.0
  perspective: 0.0
  flipud: 0.5
  fliplr: 0.5
  mosaic: 1.0
  mixup: 0.0
  copy_paste: 0.0

hyperparameters:
  optimizer: 'auto'
  cos_lr: False
  lr0: 0.01
  lrf: 0.01
  momentum: 0.937
  weight_decay: 0.0005
  warmup_epochs: 3.0
  warmup_momentum: 0.8
  warmup_bias_lr: 0.1
  dropout: 0.0

However, at least the augmentation parameters are ignored and the default values are used. I can see that in the args.yaml file in the train folder after finish a training. I tried to set hsv_h, hsv_s, and hsv_v to 0. Here is an excerpt from the args.yaml file, where these values are set to default values:

hsv_h: 0.015
hsv_s: 0.7
hsv_v: 0.4

Do you have an idea what I am doing wrong?

**Comment by glenn-jocher** [comment]
Thank you for your detailed message. Currently, augmentation parameters need to be passed directly in the `model.train()` function rather than in the `config.yaml` file. Please update your training script to include these parameters directly. If the issue persists, ensure you are using the latest version of the Ultralytics package.

**Comment by BitMarkus** [comment]
Sorry for my late reply, but I just came back from holidays. Do you mean that I should pass every single parameter seperatly to the train function? Or do you mean that I can put all train configurations in one yaml file and reference in the train function to it?
I am confused of why these things are changing so quickly.

**Comment by glenn-jocher** [comment]
You should pass augmentation parameters directly in the `train()` function. Currently, this is the recommended approach for ensuring they are applied correctly.

**Comment by cospectrum** [comment]
@BitMarkus 
Try to pass path of your `yml` augmentation config as an argument: `train(..., cfg=cfg_path)`, this overrides default params for me.
My `override.yml` config, for example:
```yml
hsv_h: 0.015              # 0.0 - 1.0    Adjusts the hue of the image by a fraction of the color wheel, introducing color variability. Helps the model generalize across different lighting conditions.
hsv_s: 0.7                # 0.0 - 1.0    Alters the saturation of the image by a fraction, affecting the intensity of colors. Useful for simulating different environmental conditions.
hsv_v: 0.4                # 0.0 - 1.0    Modifies the value (brightness) of the image by a fraction, helping the model to perform well under various lighting conditions.
degrees: 0.0              # -180 - +180  Rotates the image randomly within the specified degree range, improving the model's ability to recognize objects at various orientations.
translate: 0.1            # 0.0 - 1.0    Translates the image horizontally and vertically by a fraction of the image size, aiding in learning to detect partially visible objects.
scale: 0.2                # >=0.0        Scales the image by a gain factor, simulating objects at different distances from the camera.
shear: 0.0                # -180 - +180  Shears the image by a specified degree, mimicking the effect of objects being viewed from different angles.
perspective: 0.0          # 0.0 - 0.001  Applies a random perspective transformation to the image, enhancing the model's ability to understand objects in 3D space.
flipud: 0.0               # 0.0 - 1.0    Flips the image upside down with the specified probability, increasing the data variability without affecting the object's characteristics.
fliplr: 0.0               # 0.0 - 1.0    Flips the image left to right with the specified probability, useful for learning symmetrical objects and increasing dataset diversity.
bgr: 0.0                  # 0.0 - 1.0    Flips the image channels from RGB to BGR with the specified probability, useful for increasing robustness to incorrect channel ordering.
mosaic: 0.13              # 0.0 - 1.0    Combines four training images into one, simulating different scene compositions and object interactions. Highly effective for complex scene understanding.
mixup: 0.0                # 0.0 - 1.0    Blends two images and their labels, creating a composite image. Enhances the model's ability to generalize by introducing label noise and visual variability.
copy_paste: 0.0           # 0.0 - 1.0    Copies objects from one image and pastes them onto another, useful for increasing object instances and learning object occlusion.
auto_augment: randaugment #    str       Automatically applies a predefined augmentation policy (randaugment, autoaugment, augmix), optimizing for classification tasks by diversifying the visual features.
erasing: 0.0              # 0.0 - 0.9    Randomly erases a portion of the image during classification training, encouraging the model to focus on less obvious features for recognition.
crop_fraction: 0.1        # 0.1 - 1.0    Crops the classification image to a fraction of its size to emphasize central features and adapt to object scales, reducing background distractions.
```

**Comment by BitMarkus** [comment]
Thank you cospectrum! I will give it a try. Currently, I put all of these settings in the train function, which also works. But a seperate file would be more convenient.

**Comment by glenn-jocher** [comment]
You're welcome! Using a separate file for configurations can indeed streamline your workflow. If you have any more questions, feel free to ask.

**Comment by EconML** [comment]
I'm facing a similar issue. I'm aware we can pass the .yaml directly to the train method, but what if I am working inside the yolov5 docker container and want to pass the custom hyps using CLI?

**Comment by glenn-jocher** [comment]
You can pass custom hyperparameters via CLI in the YOLOv5 Docker container by using the `--hyp` flag, pointing to your custom `.yaml` file, like this: `python train.py --data data.yaml --hyp custom_hyps.yaml`.

**Comment by EconML** [comment]
@glenn-jocher thank you for the response. However, I was referring to running something like `yolo detect train` and wondering how to pass custom hyps in that way. I tried defining them in the `.yaml` file provided to the `data` cli arg, but they were not recognized. I resulted to just putting all of the options listed [here](https://docs.ultralytics.com/modes/train/) inside a `.yaml` file that I dynamically parse to generate the cli launch command.

**Comment by glenn-jocher** [comment]
To pass custom hyperparameters via CLI, use the `--hyp` argument to specify a separate `.yaml` file containing your hyperparameters. Make sure your `.yaml` file is structured correctly for YOLO to recognize the parameters. For example: `yolo detect train data=data.yaml model=yolo11n.pt hyp=custom_hyps.yaml`. More details can be found in the [Train documentation](https://docs.ultralytics.com/modes/train/).
